lib/open_api_spex/open_api/map.ex

```diff
@@ -0,0 +1,395 @@
+defmodule OpenApiSpex.OpenApi.Map do
+  @moduledoc """
+  This module exposes functionality to convert an arbitrary map into a OpenApi struct.
+  """
+  alias OpenApiSpex.{
+    Components,
+    Contact,
+    Discriminator,
+    Encoding,
+    Example,
+    ExternalDocumentation,
+    Header,
+    Info,
+    License,
+    Link,
+    MediaType,
+    OAuthFlow,
+    OAuthFlows,
+    OpenApi,
+    Operation,
+    Parameter,
+    PathItem,
+    Reference,
+    RequestBody,
+    Response,
+    Schema,
+    SecurityScheme,
+    Server,
+    ServerVariable,
+    Tag,
+    Xml
+  }
+
+  def decode(%{openapi: _openapi, info: _info, paths: _paths} = map) do
+    map
+    |> to_struct(OpenApi)
+    |> to_struct(:info, Info)
+    |> to_struct(:paths, PathItems)
+    |> to_struct(:servers, Servers)
+    |> to_struct(:components, Components)
+    |> to_struct(:tags, Tag)
+    |> to_struct(:externalDocs, ExternalDocumentation)
+  end
+
+  defp embedded_ref_or_struct(list, mod) when is_list(list) do
+    list
+    |> Enum.map(fn
+      %{"$ref": _} = v -> struct(Reference, v)
+      v -> to_struct(v, mod)
+    end)
+  end
+
+  defp embedded_ref_or_struct(map, mod) when is_map(map) do
+    map
+    |> Map.new(fn
+      {k, %{"$ref": _} = v} ->
+        {k, struct(Reference, v)}
+
+      {k, v} ->
+        {k, to_struct(v, mod)}
+    end)
+  end
+
+  defp convert_value_to_atom_if_present(map, key) do
+    if Map.has_key?(map, key) do
+      Map.update!(map, key, fn
+        nil ->
+          nil
+
+        v ->
+          String.to_atom(v)
+      end)
+    else
+      map
+    end
+  end
+
+  defp prepare_schema(map) do
+    map
+    |> convert_value_to_atom_if_present(:type)
+    |> to_list_of_atoms(:required)
+  end
+
+  defp to_list_of_atoms(map, key) do
+    if Map.has_key?(map, key) do
+      Map.update!(map, key, fn
+        nil ->
+          nil
+
+        v ->
+          v
+          |> Enum.map(&String.to_atom/1)
```

Comment:
Also, is it really safe to convert all of them to atoms? Simply converting all keys to atoms can open up Denial of Service vulnerabilities. It is usually better to check the value of a string against a list of safe atoms or use `String.to_existing_atom`.

Reply:
This is a problem with parsing OpenAPI documents that I can't find a solution to. The only thing I can think of is to add something to the documentation that warns against accepting OpenAPI documents from untrusted users.

Maybe there's a way to keep some values as strings, and OpenApiSpex can be modified to work with either strings or atoms for these values.

Reply:
Good point, we should document this as pointed out by @moxley. I'm happy to add that to this pull-request.

> Maybe there's a way to keep some values as strings, and OpenApiSpex can be modified to work with either strings or atoms for these values.

I'm almost certain there is a way, do we think it's worth increasing the scope of this pull-request to address the above?

Reply:
I think the scope of this PR is big enough to defer other things for later. To summarize the problem as I see it, using the `:object` schema as an example, OpenApiSpec uses atoms for the property names in the `properties:` map. Since the decoder doesn't know the property names ahead of time, it's forced to call `Atom.to_string/1` on them. We could modify OpenApiSpex to use string keys for property names, and that would solve that problem. The same thing could be applied for the other places strings are converted to atoms, such as enums. I'd like to hear your ideas on this.

There may be places in the decoder where `Atom.to_string/1` calls can be replaced with `Atom.to_existing_atom/1` without making changes to the existing OpenApiSpex. For example for `Parameter` aren't the possible values for the `in` attribute known ahead of time?

Reply:
👍 Let's document this behaviour in the README, recommending that users only load their API spec from a known file at application startup, not dynamically from external sources at runtime until we support string keyed property names.

Reply:
Agreed!